src/pop_manager.erl

```diff
@@ -0,0 +1,318 @@
+    end.
+
+setup_range_request(UnknownBlock, PopManager) ->
+    PC = PopManager#pop_manager.pop_chain,
+    N = PopManager#pop_manager.config#pop_manager_config.request_range_backup,
+
+    KnownBlock1 = pop_chain:get_head_block(PC),
+    KnownBlock2 = get_nth_prev_block(N, KnownBlock1, PC),
+    
+    {maps:get(this_id, KnownBlock1), maps:get(this_id, KnownBlock2), maps:get(this_id, UnknownBlock)}.
+
+%% given unknown block request, and longest known block from another party
+%% make a best guess about which blocks are unknown to the other party
+%% and return a array of hashes of those blocks
+
+compute_block_hash_range(UnknownBlockHash, KnownBlockHash, PopManager) ->
+    PC = PopManager#pop_manager.pop_chain,
+    UnknownBlock = pop_chain:find_block_by_id(UnknownBlockHash, PC),
+    KnownBlock = pop_chain:find_block_by_id(KnownBlockHash, PC),
+
+    compute_block_hash_range(UnknownBlock, KnownBlock, PC, []).
+
+compute_block_hash_range(UnknownBlock, KnownBlock, PC, AccList) ->
+    UnknownId = maps:get(this_id, UnknownBlock),
+    KnownId = maps:get(this_id, KnownBlock),
+
+    UnknownHeight = maps:get(height, UnknownBlock),
+    KnownHeight = maps:get(height, KnownBlock),
+
+    if UnknownId == KnownId ->
+	    AccList;  % done, found common parent
+
+       UnknownId /= KnownId ->
+	    %% decrement higher block, or both if equal height
+	    %% each time Unknown is decremented, add its hash to the list
+
+	    if UnknownHeight >= KnownHeight -> 	
+		    UnknownBlockNew = get_prev(UnknownBlock, PC), 
+	            KnownBlockNew = KnownBlock,
+	            AccListNew = [UnknownId | AccList];
+
+	       UnknownHeight =< KnownHeight ->
+		    UnknownBlockNew = UnknownBlock, 
+	            KnownBlockNew = get_prev(KnownBlock, PC),
+	            AccListNew = AccList;
+
+	       UnknownHeight == KnownHeight ->
+		    UnknownBlockNew = get_prev(UnknownBlock, PC), 
+	            KnownBlockNew = get_prev(KnownBlock, PC),
+	            AccListNew = [UnknownId | AccList]
+	    end,
+
+	    compute_block_hash_range(UnknownBlockNew, KnownBlockNew, PC, AccListNew)
+    end.
+    
+%% Adds a new block into pop_chain.
+%% Then it checks unbound blocks, and adds all that become bound.
+%% Function hook on_new_block is triggered on each addition.
+
+add_block_in_order(Block, CurrentTime, PopManager0) ->
+    PC0 = PopManager0#pop_manager.pop_chain,
+    UnboundMap0 = PopManager0#pop_manager.unbound_blocks,
+
+    OnNewBlockFn = PopManager0#pop_manager.config#pop_manager_config.on_new_block,
+
+    PC1 = pop_chain:add_block_in_order(Block, CurrentTime, PC0),
+
+    OnNewBlockFn(Block),
+
+    %% process unbound blocks, recursively add new ones as needed
+
+    IsUnboundPred = 
+	fun({_Id, B}) ->
+		R = pop_chain:find_block_by_id(maps:get(previous_id, B), PC1),
+		case R of
+		    {ok, _} ->
+			false;
+		    error ->
+			true
+		end
+	end,
+
+    {UnboundList, BoundList} = lists:partition(IsUnboundPred, maps:to_list(UnboundMap0)),
```

Comment:
:o

Reply:
👍

Reply:
Such standard library use instead of monster foldls

Annoyed though maps doesn't have partition function